process/headerCheck/headerVersioningHandler.go

```diff
@@ -0,0 +1,205 @@
+	})
+
+	currentEpoch := uint32(0)
+	for _, ver := range versionsByEpochs {
+		if ver.StartEpoch >= ver.EndEpoch {
+			return nil, fmt.Errorf("%w, StartEpoch is greater or equal to EndEpoch, version %s",
+				ErrInvalidVersionOnEpochValues, ver.Version)
+		}
+		if currentEpoch != ver.StartEpoch {
+			return nil, fmt.Errorf("%w for version %s",
+				ErrInvalidVersionFoundGaps, ver.Version)
+		}
+		if len(ver.Version) > core.MaxSoftwareVersionLengthInBytes {
+			return nil, fmt.Errorf("%w for version %s",
+				ErrInvalidVersionStringTooLong, ver.Version)
+		}
+
+		currentEpoch = ver.EndEpoch
+	}
+
+	return versionsByEpochs, nil
+}
+
+// GetVersion returns the version by providing the epoch
+func (hvh *headerVersioningHandler) GetVersion(epoch uint32) string {
+	ver := hvh.getMatchingVersion(epoch)
+	if ver == wildcard {
+		return hvh.defaultVersion
+	}
+
+	return ver
+}
+
+// GetVersion returns the version by providing the epoch
+func (hvh *headerVersioningHandler) getMatchingVersion(epoch uint32) string {
+	storedVersion, ok := hvh.getFromCache(epoch)
+	if ok {
+		return storedVersion
+	}
+
+	for _, ver := range hvh.versions {
+		if ver.StartEpoch <= epoch && epoch < ver.EndEpoch {
+			hvh.setInCache(epoch, ver.Version)
+
+			return ver.Version
+		}
+	}
+
+	log.Debug("headerVersioningHandler.GetVersion version not found",
+		"epoch", epoch, "default version", hvh.defaultVersion)
+
+	hvh.setInCache(epoch, hvh.defaultVersion)
+
+	return hvh.defaultVersion
+}
+
+func (hvh *headerVersioningHandler) getFromCache(epoch uint32) (string, bool) {
+	key := make([]byte, keySize)
+	binary.BigEndian.PutUint32(key, epoch)
+
+	obj, ok := hvh.versionCache.Get(key)
+	if !ok {
+		return "", false
+	}
+
+	str, ok := obj.(string)
+
+	return str, ok
+}
+
+func (hvh *headerVersioningHandler) setInCache(epoch uint32, version string) {
+	key := make([]byte, keySize)
+	binary.BigEndian.PutUint32(key, epoch)
+
+	_ = hvh.versionCache.Put(key, version, len(key)+len(version))
+}
+
+// Verify will check the header's fields such as the chain ID or the software version
+func (hvh *headerVersioningHandler) Verify(hdr data.HeaderHandler) error {
+	if len(hdr.GetReserved()) > 0 {
+		return process.ErrReservedFieldNotSupportedYet
+	}
+
+	err := hvh.checkSoftwareVersion(hdr)
+	if err != nil {
+		return err
+	}
+
+	return hvh.checkChainID(hdr)
+}
+
+func (hvh *headerVersioningHandler) checkVersionLength(version []byte) error {
+	if len(version) == 0 || len(version) > core.MaxSoftwareVersionLengthInBytes {
+		return fmt.Errorf("%w when checking lenghts", ErrInvalidSoftwareVersion)
+	}
+
+	return nil
+}
+
+// checkSoftwareVersion returns nil if the software version has the correct length
+func (hvh *headerVersioningHandler) checkSoftwareVersion(hdr data.HeaderHandler) error {
+	err := hvh.checkVersionLength(hdr.GetSoftwareVersion())
+	if err != nil {
+		return err
+	}
+
+	version := hvh.getMatchingVersion(hdr.GetEpoch())
+	if version == wildcard {
```

Comment:
This shouldn't accept anything, it should accept the default version only

Reply:
we can make use of wildcards, if for example, we want to change the default string. We temporarily put the next versions on wildcard, change what we want (during this change *everything* will be accepted) and then put the desired string.

Reply:
ok